From 3b6f3930102d3616fc68cd0a5833423d1cc3344d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Timo=20Ter=C3=A4s?= <timo.teras@iki.fi>
Date: Fri, 1 Nov 2024 12:38:57 +0200
Subject: [PATCH] hash: remove bad cast of delete_item to enumerator callback

(cherry picked from commit a9ef3480df7a9264bfc7076b0bcf5891ae7f36ab)
---
 src/hash.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/hash.c b/src/hash.c
index bf8041deb2e..e8eed575cf6 100644
--- a/src/hash.c
+++ b/src/hash.c
@@ -19,9 +19,15 @@ void apk_hash_init(struct apk_hash *h, const struct apk_hash_ops *ops,
 	h->num_items = 0;
 }
 
+static int apk_hash_free_item_enumerator(apk_hash_item item, void *ctx)
+{
+	((apk_hash_delete_f) ctx)(item);
+	return 0;
+}
+
 void apk_hash_free(struct apk_hash *h)
 {
-	apk_hash_foreach(h, (apk_hash_enumerator_f) h->ops->delete_item, NULL);
+	if (h->ops->delete_item) apk_hash_foreach(h, apk_hash_free_item_enumerator, h->ops->delete_item);
 	apk_hash_array_free(&h->buckets);
 }
 
@@ -35,8 +41,7 @@ int apk_hash_foreach(struct apk_hash *h, apk_hash_enumerator_f e, void *ctx)
 	foreach_array_item(bucket, h->buckets) {
 		hlist_for_each_safe(pos, n, bucket) {
 			r = e(((void *) pos) - offset, ctx);
-			if (r != 0 && ctx != NULL)
-				return r;
+			if (r != 0) return r;
 		}
 	}
 
-- 
2.47.1

